modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1249 @@
+        try! {
+          _split(_tail!);
+        }
+      }
+    }
+
+    pragma "no doc"
+    proc _append(x: eltType)
+    lifetime this < x { 
+      _size += 1;
+      _spareSpaceInTail();
+      _tail!.append(x);
+    }
+
+    pragma "no doc"
+    proc _append(ref x: eltType) where isOwnedClass(x)
+    lifetime this < x { 
+      _size += 1;
+      _spareSpaceInTail();
+      _tail!.append(x);
+    }
+
+    /*
+      Add an element to the end of this unrolledLinkedList.
+
+      :arg x: An element to append.
+      :type x: `eltType`
+    */
+    proc ref append(x: eltType)
+    lifetime this < x {
+      _enter();
+      _append(x);
+      _leave();
+    }
+
+    pragma "no doc"
+    proc append(ref x: eltType) where isOwnedClass(x)
+    lifetime this < x {
+      _enter();
+      _append(x);
+      _leave();
+    }
+
+    /*
+      Returns `true` if this unrolledLinkedList contains an element equal to the value of
+      `x`, and `false` otherwise.
+
+      :arg x: An element to search for.
+      :type x: `eltType`
+
+      :return: `true` if this unrolledLinkedList contains `x`.
+      :rtype: `bool`
+    */
+    proc const contains(x: eltType): bool {
+      var result = false;
+
+      on this {
+        _enter();
+
+        for item in this do
+          if item == x {
+            result = true;
+            break;
+          }
+
+        _leave();
+      }
+
+      return result;
+    }
+
+    /*
+      Returns a reference to the first item in this unrolledLinkedList.
+
+      .. warning::
+
+        Calling this method on an empty unrolledLinkedList will cause the currently running
+        program to halt. If the `--fast` flag is used, no safety checks will
+        be performed.
+
+      :return: A reference to the first item in this unrolledLinkedList.
+      :rtype: `ref eltType`
+    */
+    proc ref first() ref {
+      _enter();
+
+      if boundsChecking && _size == 0 {
+        _leave();
+        boundsCheckHalt("Called \"unrolledLinkedList.first\" on an empty unrolledLinkedList.");
+      }
+
+      ref result = _getRef(0);
+      _leave();
+
+      return result;
+    }
+
+    /*
+      Returns a reference to the last item in this unrolledLinkedList.
+
+      .. warning::
+
+        Calling this method on an empty unrolledLinkedList will cause the currently running
+        program to halt. If the `--fast` flag is used, no safety checks will
+        be performed.
+
+      :return: A reference to the last item in this unrolledLinkedList.
+      :rtype: `ref eltType`
+    */
+    proc ref last() ref {
+      _enter();
+
+      if boundsChecking && _size == 0 {
+        _leave();
+        boundsCheckHalt("Called \"unrolledLinkedList.last\" on an empty unrolledLinkedList.");
+      }
+     
+      ref result = _tail!.data[_tail!.size-1];
+      _leave();
+
+      return result;  
+    }
+
+    pragma "no doc"
+    inline proc ref _extendGeneric(collection) {
+      on this {
+        for item in collection {
```

Comment:
If you have any optimization ideas for this, you can add them here as `TODO` comments.